### PR TITLE
Us-15 multistep navigation

### PR DESCRIPTION
**Pull request by shanvinluo**
# MultistepNavigation Feature Implementation
This PR closes issue #7 MultistepNavigation feature that enables users to navigate between and within buildings using a step-by-step approach.

## Key Features
- Transition between outdoor directions and indoor room-to-room navigation
- Progress through navigation steps with Next/Previous controls
- Set origin and destination by location, building, or specific room
- Shows which floors to navigate between with clear visual indicators
- Works with both Google Maps for outdoor navigation and floor plans for indoor navigation

## Technical Implementation
Navigation Strategy Pattern: Modular design using strategy pattern to handle different navigation scenarios
Multi-Step Navigation Engine: Core navigation plan system that creates and manages navigation steps
Responsive Components: UI adapts to different navigation states (loading, errors, completed steps)

## Testing
Unit tests for all navigation services
Component tests for UI elements and interactions
Integration tests for the entire navigation flow
Edge case handling for various room formats, building types, and navigation scenarios
Mock implementations of external services for reliable testing

## Images

| Screenshot 2025-04-05 16:29:29 | Screenshot 2025-04-05 16:31:38 |
| Screenshot 2025-04-05 16:31:42 | Screenshot 2025-04-05 16:31:52 |
| Screenshot 2025-04-05 16:31:58 | &nbsp;

### CODE REVIEW

**Comment by JeremyVieira** [comment]
Fixed the merged conflicts and failing tests. Please review the code thoroughly just to make sure everything is working. Did the best I can and couldn't find any bugs and everything is working normally

File: ConcordiaMaps/components/OutdoorNavigation/DirectionsBox.js [hunk 1]
```diff
@@ -1,53 +1,32 @@
-import React, { useState, useEffect, useRef } from "react";
+import React, { useEffect, useState } from "react";
```
Comment (GitExcited) [inline ConcordiaMaps/components/OutdoorNavigation/DirectionsBox.js:1]: The useRef import has been removed, will that cause issues in terms of code functionality?

File: ConcordiaMaps/components/OutdoorNavigation/ExpandedMapModal.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import React from "react";
+import { View, Text, TouchableOpacity } from "react-native";
+import { WebView } from "react-native-webview";
+import PropTypes from "prop-types";
+import MapGenerationService from "../../services/MapGenerationService";
+
+/**
```
Comment (GitExcited) [inline ConcordiaMaps/components/OutdoorNavigation/ExpandedMapModal.js:7]: This file is newly added but lacks comments and documentation. Can you provide comments explaining the key functionalities and any important considerations for future maintenance?

File: ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js [hunk 3]
```diff
@@ -37,8 +37,8 @@ const FloatingSearchBar = ({
 
       // Convert to base64 string
       let base64 = "";
-      for (let i = 0; i < randomBytes.length; i++) {
-        base64 += String.fromCharCode(randomBytes[i]);
+      for (const byte of randomBytes) {
```
Comment (GitExcited) [inline ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js:40]: I like this refactor of the loop to a `foor .. of` loop. However, the error handling for the expo-crypto import could be more robust. What happens if the import fails? Should we add a fallback mechanism or error logging?

File: ConcordiaMaps/components/OutdoorNavigation/GetDirections.js [hunk 4]
```diff
@@ -111,17 +106,6 @@ const GetDirections = () => {
     }
   }, [disableLiveLocation]);
 
-  useEffect(() => {
-    if (fromPopup) {
```
Comment (GitExcited) [inline ConcordiaMaps/components/OutdoorNavigation/GetDirections.js:115]: The changes of the organization of this file are good. However, can you explain why the removal of the fromPopup? Its effects seem important

File: ConcordiaMaps/components/userInPolygon.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import { Building } from "../data/markersData";
+import { useState, useContext, useEffect } from "react";
+import { LocationContext } from "../contexts/LocationContext";
+/*
+this function determines whether the user is inside
+one of the known to use buildings: points of interests or Concordia buildings
+if a user is inside a concordia building and they are willing to go outside,
+we will generate the internal navigation card to leave the building.
+Such option won't be available for exiting the external point of interest. 
+We will simply suggest "navigate to the exit of {current_building}"
```
Comment (GitExcited) [inline ConcordiaMaps/components/userInPolygon.js:12]: This utility function is newly added and has great Documentation. It allows us to understand its usage and will be helpful for future maintainability and scalability

**Review by GitExcited — CHANGES_REQUESTED** [review]
This file has a lot of changes. However, most of them were address in PR #200 Therefore, I only asked some extra clarifications on certain files.

File: ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js (hunk 3 above)
Comment (JeremyVieira) [reply ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js:40]: The fromPopup was replaced with a useEffect so that it automatically sets the destination which fromPopup didnt do

File: ConcordiaMaps/components/OutdoorNavigation/GetDirections.js (hunk 4 above)
Comment (JeremyVieira) [reply ConcordiaMaps/components/OutdoorNavigation/GetDirections.js:115]: The fromPopup was replaced with a useEffect so that it automatically sets the destination which fromPopup didnt do

File: ConcordiaMaps/components/OutdoorNavigation/ExpandedMapModal.js (hunk 2 above)
Comment (JeremyVieira) [reply ConcordiaMaps/components/OutdoorNavigation/ExpandedMapModal.js:7]: added some comments and documentation to make it easier for future use

File: ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js (hunk 3 above)
Comment (JeremyVieira) [reply ConcordiaMaps/components/OutdoorNavigation/FloatingSearchBar.js]: removed the token handling from this file and moved the responsibility to the useGoogleMapDirections hook which has better implementation and fallbacks

**Review by GitExcited — APPROVED** [review]
Thank you for making the changes I requested. The PR is ready to merge

File: ConcordiaMaps/components/CalendarScreen.js [hunk 6]
```diff
@@ -195,10 +187,11 @@ const CalendarScreen = () => {
                 <TouchableOpacity
                   testID="getClassDirectionsButton"
                   style={styles.classDirectionsButton}
-                  onPress={() => handleGetDirections(item)}
+                  onPress={() => getDirectionsTo(item.location)}
                 >
                   <Text style={styles.classDirectionsButtonText}>
-                    Get Directions
```
Comment (MomoG221) [inline ConcordiaMaps/components/CalendarScreen.js:201]: The 'Get Directions' button isn’t working after the new updates. It just opens the external navigation menu instead of going to Multitipse navigation.

**Review by MomoG221 — CHANGES_REQUESTED** [review]
The rest of the functionality is working well. I already gave some refactoring comments in the previous pull request, which was merged into this one. Just fix the issue with the 'Get Directions' button from the calendar screen, and it’ll be ready to merge.

File: ConcordiaMaps/components/CalendarScreen.js (hunk 6 above)
Comment (MomoG221) [reply ConcordiaMaps/components/CalendarScreen.js:201]: Was fixed here https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/pull/239/commits/e91467c394decb592c0ac33ed53e245aa1bcc8e7

**Review by MomoG221 — APPROVED** [review]
